mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -141,19 +168,20 @@ impl PutObjectRequest for S3PutObjectRequest {
     type ClientError = S3RequestError;
 
     async fn write(&mut self, slice: &[u8]) -> ObjectClientResult<(), PutObjectError, Self::ClientError> {
-        // Check if the request has already finished (which can only happen because of an error in
-        // the request), and fail the write if so. Ordering doesn't matter here as it should be
-        // impossible for the request to succeed while we still hold `&mut self`, as no one can call
-        // `complete()`.
-        select_biased! {
-            result = self.writer.write(slice).fuse() => {
-                self.total_bytes += slice.len() as u64;
-                result.map_err(|e| S3RequestError::InternalError(Box::new(e)).into())
-            }
-
-            // Request can't have succeeded if we still own `&mut self`
-            result = (&mut self.body).fuse() => Err(result.expect_err("request can't succeed while still writing")),
+        // On first write, check the pending CreateMultipartUpload.
+        if let Some(create_mpu) = self.pending_create_mpu.take() {
+            // Wait for CreateMultipartUpload to complete successfully, or the MPU to fail.
+            create_mpu.await.unwrap()?;
         }
+
+        // Write will fail if the request has already finished (because of an error).
+        self.body
+            .meta_request
+            .write(slice, false)
```

Comment:
Good spot. Addressed now by having `MetaRequest::write` (and `MetaRequestWrite`) holding exclusive access to the meta-request until completion.

However, in practice, the same guarantee already held because we only called it through `S3PutObjectRequest::write` (which also keeps exclusive access).